Identification try2

picobyte opened the pull request.

Grokmoo commented on sulis_core/src/util.rs:
```diff
@@ -110,6 +111,16 @@ pub fn gen_rand<T: SampleUniform + Sized>(min: T, max: T) -> T {
     rand::thread_rng().gen_range(min, max)
 }
 
+pub fn gen_rand_weight<'a, T, V>(items: &'a [(T, V)]) -> &'a T
```
This is displaying quite different behavior than the old.  It seems to be completely ignoring "none" as a value for an adjective.  For example, loot list level1_weapons always generates an adjective2 whereas it should only do less than 50% of the time for most weapons.

picobyte replied:
You mean you observe this change? strange, the "none" check is in the caller which is in loot_list(), will look into it..

Grokmoo replied:
If I understand what you are intending correctly, the level 1 weapons and armor loot lists should only have a 5% chance of generating an unidentified item based on the data?  But, in my play testing, they seem to generate one every time.  And the behavior is similar where "none" should be the most common for the other adjectives but you never see none.  Your implementation seems to be skipping the probability of the first entry in the adjectives list.

Grokmoo commented on data/loot_lists/level10/level10_armor.yml:
```diff
@@ -13,6 +13,7 @@ weighted_entries:
       exceptional: 5
       fine: 5
     adjective2:
+      unidentified: 12
```
is this mutually exclusive with the other adjective2 entries?  Shouldn't it be its own adjective3 entry?

picobyte replied:
We won't ever allow multiple bonuses? I added it here and implemented the re-roll in the assumption that we might want to allow multiple bonuses in the future.

Grokmoo commented on sulis_module/src/inventory_builder.rs:
```diff
@@ -142,7 +142,7 @@ impl InventoryBuilder {
                 }
                 Some(item) => {
-                    let state = ItemState::new(item, entry.item.variant);
+                    let state = ItemState::new(item, entry.item.variant, true);
```
why isn't unidentified configurable in this context?  (i.e. why is it always true)

picobyte replied:
I thought this generated player character items at the start, do we ever want to add a pc that has a cursed item from start? Oh well, it could be an interesting start perspective, so added.

Grokmoo commented on sulis_module/src/item.rs:
```diff
@@ -132,18 +132,7 @@ impl Item {
             &all_adjectives,
         );
 
-        let mut name = String::new();
```
Removing the precompute of the name and computing it every time `get_name()` is called seems like it could be a pretty big performance hit.  Keep in mind this method could potentially be called hundreds (thousands?) of times on every UI state change.

I'd suggest keeping the name here and maybe also storing an unidentified name variant.

picobyte replied:
ok that option crossed my mind as well, will do that.

Grokmoo commented on sulis_module/src/loot_list.rs:
```diff
@@ -301,47 +285,37 @@ impl LootList {
+            while remain > 0 {
+                let pick = gen_rand_weight(&entry.adjective2);
+                match pick.as_str() {
+                    "unidentified" => {
```
i do not like how "unidentified" is hard coded here to only appear in adjective2.  it shouldn't need special treatment here, if it were to just appear in its own regular, mutually exclusive category.  such as adjective3.

Really, adjectives should be an array of arbitrary length.  But that doesn't need to be fixed in this PR.

picobyte replied:
I have it fixed anyway. adjective[123] -> adjectives.

Grokmoo commented on sulis_state/src/item_list.rs:
```diff
@@ -108,6 +108,14 @@ impl ItemList {
         Some(self.items.remove(index))
     }
 
+    pub fn identify(&mut self, index: usize) -> Option<&ItemState> {
```
This approach that you have taken does not store the identified status in the save.  So identifying an item does not persist when saving.  (You need to re-identify every time you save and load).

picobyte replied:
should be resolved after last update

Grokmoo commented on sulis_state/src/merchant_state.rs:
```diff
@@ -40,19 +41,21 @@ impl MerchantState {
             loot_list_id: save.loot_list_id,
             buy_frac: save.buy_frac,
             sell_frac: save.sell_frac,
+            identify_fraction: 0.2,
```
identify_fraction should not be hardcoded to 0.2.  Merchants are currently defined ad-hoc in conversation files, for example `campaigns/twin_expanse_act1/conversations/dwarf_smith.yml`.

You could `#[serde_default]` this to 0.2 as a default, but it should be changable in the `MerchantData` in `sulis_module/on_trigger`, and it should save in the MerchantSaveState.

picobyte replied:
Fixed

Grokmoo commented on sulis_state/src/merchant_state.rs:
```diff
@@ -134,7 +138,12 @@ impl MerchantState {
+        let value = if item_state.identified {
+            item_state.item.value as f32
+        } else {
+            (1.0 - self.identify_fraction) * item_state.item.value.abs() as f32
```
I'm not sure the player should really know the value of an identified item before it is identified.  I think in a lot of games an unidentified item sells for a nominal amount, for example the value of the base item without any bonuses.

But I think this is probably ok as-is for a first pass.

picobyte replied:
Agreed.

Grokmoo commented on sulis_module/src/loot_list.rs:
```diff
@@ -262,8 +245,9 @@ impl LootList {
                     }
                     Some(item) => item,
                 };
+                let identified = !adjectives.contains(&"unidentified".to_string());
```
You should just be able to use `"unidentified"` directly here rather than `&"unidentified".to_string()`.  This is allocating memory to create a new string and then returning a reference to it, which is unnecessary.

However, `"unidentified"` should really be configurable as you have declared an unidentified item_adjective above.    I would put `unidentified_item_adjective` as a rule in rules.rs.

picobyte replied:
I believe I needed this because "unidentified" is an &str, and this requires an &String. I'll try the rule.

Grokmoo commented on sulis_view/src/item_button.rs:
```diff
@@ -258,15 +258,21 @@ impl ItemButton {
-            Some(action)
+            let merchant = Widget::kind_mut::<MerchantWindow>(&window_widget).player();
+
+            if self.get_item_state().filter(|i| !i.identified).is_some() {
```
I am ok with changing the sell button to identified for a first pass.  But there is no feedback to the user at all about how much the identification will cost.

picobyte replied:
I see, to be storing "unidentified" several times in adjectives, which causes problems.

picobyte replied:
this is fixed

Grokmoo commented on sulis_view/src/item_callback_handler.rs:
```diff
@@ -174,6 +174,36 @@ pub fn sell_item_cb(entity: &Rc<RefCell<EntityState>>, index: usize) -> Callback
+            if let Some(item_state) = stash.borrow_mut().identify(index) {
+                let price = merchant.get_sell_price(&item_state) as f32;
+                let charge = price * merchant.identify_fraction;
+                GameState::add_party_coins(-charge.ceil() as i32);
```
you need to validate that the party actually has enough coins to afford the identification before doing this.

picobyte replied:
good catch

picobyte commented:
> I think you missed the level1 loot_lists? Or was that intentional?
Not sure anymore, but added.

Several issues are addressed in latest pull, but some more changes will follow, these should probably include:

unreplied reported issues
Accessories could have unidentified status, although it's adjective 2.
multiple "unidentified" adjectives in save for items

picobyte commented:
Hope all is well, life somewhat grinds to a halt here, in Europe. I've addressed most issues, not sure what's causes the weighted selection difference, and there are some open questions

Grokmoo commented:
With the kids at home all day and still working full time, I am bit limited on free time at the moment.

I'll try to take a look at your recent changes soon.  Hope you are doing well.

picobyte commented:
That's good to hear, don't worry about this pull request, priorities first stay safe and take care.

Grokmoo commented on data/loot_lists/level1/level1_accessories.yml:
```diff
@@ -15,10 +15,12 @@ weighted_entries:
       warlock_power1: 2
       mage_ability1: 2
       ability1: 2
+    - identified: 90
```
Shouldn't we just be using "none" here?  do we need the separate "identified" item adjective?

Grokmoo commented on data/loot_lists/level1/level1_accessories.yml:
```diff
@@ -5,8 +5,8 @@ generate:
 weighted_entries:
   amulet_gold:
     weight: 1
-    adjective1:
```
Thank you for updating all the loot lists to an array format; that is quite a bit of work!

Grokmoo commented on sulis_module/src/item.rs:
```diff
@@ -132,15 +133,24 @@ impl Item {
         for adj in all_adjectives.iter() {
             if let Some(ref prefix) = adj.name_prefix {
+                if unidentified_name.is_none() &&
+                    "Exceptional Fine Masterwork Rusty Worn ".find(prefix).is_none() {
```
The data driven values from these adjectives shouldn't be hardcoded here.  The "Unidentified" string shouldn't either.

We should use the `unidentified_item_adjective` from the `Rules`.  We should just be able to set a name_prefix for that adjective and append it to the base item name.
